File: docs/historian/historian.py
#!/usr/bin/env python3
"""OUTRAM PARK "historian" — pre-merge-to-main release report generator.

WHAT THIS PRODUCES
------------------
A single markdown file summarising a window of `develop` history before it is
merged into `main`: the **API tokens spent** and the **lines / KLOC written**
across every commit in a `DDMMYY..DDMMYY` date range, plus a per-crate breakdown
and a per-commit table.

* **Tokens** come from the `API-Usage-Since-Last-Commit:` trailers that
  `scripts/token_accounting.py` stamps into each commit message (nothing is
  estimated — commits made before the token hooks existed, or outside a Claude
  session, simply contribute 0 and are counted as "no token data").
* **Lines** come from `git log --numstat --no-merges` over the range (added /
  removed / net, with a Rust-only `.rs` figure). Merge commits are excluded so
  lines are not double-counted.

WHERE IT LIVES
--------------
Both this script and the reports it writes live under `docs/historian/` at the
workspace root. Reports are named `historian_<from>_to_<to>.md` (DDMMYY).

STANDING RULE (see workspace CLAUDE.md)
---------------------------------------
Before merging `develop` into `main`, generate the report for the window being
released and commit it alongside the merge.

USAGE
-----
    # explicit window (DDMMYY = day-month-year, 2-digit year)
    python3 docs/historian/historian.py --from 220726 --to 230726

    # default: everything on develop that is not yet on main, up to today
    python3 docs/historian/historian.py

    # a different branch / target
    python3 docs/historian/historian.py --branch develop --base main

Stdlib only.
"""
from __future__ import annotations

import argparse
import datetime as _dt
import os
import re
import subprocess
import sys

TRAILER_RE = re.compile(r"^API-Usage-Since-Last-Commit:\s*(.+)$", re.MULTILINE)
COMPONENTS = ("in", "out", "cache_read", "cache_write")
HERE = os.path.dirname(os.path.abspath(__file__))


def git(*args: str) -> str:
    try:
        return subprocess.check_output(["git", *args], stderr=subprocess.DEVNULL, text=True)
    except Exception:
        return ""


def repo_root() -> str:
    return (git("rev-parse", "--show-toplevel").strip()) or os.getcwd()


def parse_ddmmyy(s: str) -> _dt.date:
    s = s.strip()
    if not re.fullmatch(r"\d{6}", s):
        raise ValueError(f"expected DDMMYY (6 digits), got {s!r}")
    dd, mm, yy = int(s[0:2]), int(s[2:4]), int(s[4:6])
    return _dt.date(2000 + yy, mm, dd)


def ref_for_branch(branch: str) -> str:
    """Prefer the remote-tracking ref so the report reflects what is pushed."""
    for cand in (f"origin/{branch}", branch):
        if git("rev-parse", "--verify", "--quiet", cand).strip():
            return cand
    return branch


def commits_in_range(branch_ref: str, base_ref: str | None,
                     dfrom: _dt.date | None, dto: _dt.date | None) -> list[str]:
    args = ["log", "--no-merges", "--reverse", "--format=%H"]
    if dfrom:
        args.append(f"--since={dfrom.isoformat()} 00:00:00")
    if dto:
        args.append(f"--until={dto.isoformat()} 23:59:59")
    if base_ref:
        args.append(f"{base_ref}..{branch_ref}")
    else:
        args.append(branch_ref)
    out = git(*args).strip()
    return [h for h in out.splitlines() if h]


def commit_meta(sha: str) -> dict:
    FS = "\x1f"
    raw = git("show", "-s", f"--format=%h{FS}%aI{FS}%s{FS}%b", sha)
    parts = raw.split(FS)
    short = parts[0] if len(parts) > 0 else sha[:9]
    iso = parts[1] if len(parts) > 1 else ""
    subject = parts[2] if len(parts) > 2 else ""
    body = FS.join(parts[3:]) if len(parts) > 3 else ""
    toks = {c: 0 for c in COMPONENTS}
    toks["total"] = 0
    toks["has_data"] = False
    m = TRAILER_RE.search(body)
    if m:
        fields = {}
        for tok in m.group(1).split():
            if "=" in tok:
                k, v = tok.split("=", 1)
                fields[k] = v
        if fields.get("source", "none") != "none":
            toks["has_data"] = True
        for c in COMPONENTS:
            try:
                toks[c] = int(fields.get(c, 0))
            except ValueError:
                toks[c] = 0
        try:
            toks["total"] = int(fields.get("total", 0))
        except ValueError:
            toks["total"] = sum(toks[c] for c in COMPONENTS)
    return {"sha": short, "date": iso[:10], "subject": subject, "tok": toks}


def numstat(sha: str) -> tuple[int, int, int, int, dict]:
    """(added, removed, rs_added, rs_removed, per_crate_added) for one commit."""
    raw = git("show", "--numstat", "--format=", sha)
    added = removed = rs_added = rs_removed = 0
    per_crate: dict[str, int] = {}
    for line in raw.splitlines():
        line = line.strip()
        if not line:
            continue
        cols = line.split("\t")
        if len(cols) < 3:
            continue
        a, r, path = cols[0], cols[1], cols[2]
        if a == "-" or r == "-":  # binary
            continue
        # Normalise git's rename compaction to the NEW path:
        #   "crates/{old => new}/f.rs"  -> "crates/new/f.rs"
        #   "old/path/f => new/path/f"   -> "new/path/f"
        path = re.sub(r"\{[^{}]*=>\s*([^{}]*)\}", r"\1", path)
        if " => " in path:
            path = path.split(" => ", 1)[1]
        path = re.sub(r"//+", "/", path).strip()
        a_i, r_i = int(a), int(r)
        added += a_i
        removed += r_i
        if path.endswith(".rs"):
            rs_added += a_i
            rs_removed += r_i
        m = re.match(r"crates/([^/]+)/", path)
        if m:
            per_crate[m.group(1)] = per_crate.get(m.group(1), 0) + a_i
    return added, removed, rs_added, rs_removed, per_crate


def g(n: int) -> str:
    return f"{n:,}"


def render(dfrom, dto, branch_ref, base_ref, commits) -> str:
    rows = []
    tot = {c: 0 for c in COMPONENTS}
    tot["total"] = 0
    tot_added = tot_removed = tot_rs_added = tot_rs_removed = 0
    tok_covered = 0
    per_crate_total: dict[str, int] = {}
    for sha in commits:
        meta = commit_meta(sha)
        added, removed, rs_a, rs_r, pc = numstat(sha)
        for c in COMPONENTS:
            tot[c] += meta["tok"][c]
        tot["total"] += meta["tok"]["total"]
        if meta["tok"]["has_data"]:
            tok_covered += 1
        tot_added += added
        tot_removed += removed
        tot_rs_added += rs_a
        tot_rs_removed += rs_r
        for k, v in pc.items():
            per_crate_total[k] = per_crate_total.get(k, 0) + v
        rows.append((meta, added, removed))

    n = len(commits)
    span = ""
    if dfrom and dto:
        span = f"{dfrom.strftime('%d %b %Y')} → {dto.strftime('%d %b %Y')}"
    elif base_ref:
        span = f"all of `{branch_ref}` not in `{base_ref}`"
    else:
        span = f"full history of `{branch_ref}`"
    stamp = _dt.datetime.now().strftime("%Y-%m-%d %H:%M")

    out = []
    out.append(f"# OUTRAM PARK — historian report ({span})\n")
    out.append(
        "> Pre-merge-to-`main` accounting of the API tokens spent and the lines / "
        "KLOC written across this window of `develop` history. **Auto-generated** "
        f"by `docs/historian/historian.py` on {stamp}; regenerate with "
        "`python3 docs/historian/historian.py --from DDMMYY --to DDMMYY`.\n"
    )
    out.append("## Scope\n")
    out.append(f"- **Branch:** `{branch_ref}`" + (f" (vs base `{base_ref}`)" if base_ref else "") + "\n")
    out.append(f"- **Window:** {span}\n")
    out.append(f"- **Commits (non-merge):** {n}\n")
    out.append(
        f"- **Token coverage:** {tok_covered}/{n} commits carry an "
        "`API-Usage-Since-Last-Commit` trailer. Commits before the token-accounting "
        "hooks existed (or made outside a Claude session) contribute 0 and are "
        "counted here as *no token data* — that is correct, not missing data.\n"
    )
    out.append("\n## Totals\n")
    out.append("### Lines written (git numstat, merges excluded)\n")
    out.append("| Metric | Lines | KLOC |")
    out.append("|---|--:|--:|")
    out.append(f"| Added (all files) | {g(tot_added)} | {tot_added/1000:.1f} |")
    out.append(f"| Removed (all files) | {g(tot_removed)} | {tot_removed/1000:.1f} |")
    out.append(f"| **Net (all files)** | **{g(tot_added - tot_removed)}** | **{(tot_added - tot_removed)/1000:.1f}** |")
    out.append(f"| Added (Rust `.rs`) | {g(tot_rs_added)} | {tot_rs_added/1000:.1f} |")
    out.append(f"| Net (Rust `.rs`) | {g(tot_rs_added - tot_rs_removed)} | {(tot_rs_added - tot_rs_removed)/1000:.1f} |")
    out.append("\n### API tokens spent\n")
    out.append("| Component | Tokens |")
    out.append("|---|--:|")
    out.append(f"| input | {g(tot['in'])} |")
    out.append(f"| output | {g(tot['out'])} |")
    out.append(f"| cache_read | {g(tot['cache_read'])} |")
    out.append(f"| cache_write | {g(tot['cache_write'])} |")
    out.append(f"| **total** | **{g(tot['total'])}** |")
    out.append(
        "\n_`total` = input + output + cache_read + cache_write. Cache-read "
        "(prompt-cache re-reads of the growing context) usually dominates; the "
        "output figure is the closest proxy for net generated content._\n"
    )

    if per_crate_total:
        out.append("## Lines added, by crate (top 20)\n")
        out.append("| Crate | Lines added |")
        out.append("|---|--:|")
        for k, v in sorted(per_crate_total.items(), key=lambda kv: -kv[1])[:20]:
            out.append(f"| `{k}` | {g(v)} |")
        out.append("")

    out.append("## Per-commit ledger\n")
    out.append("| Date | Commit | Subject | +lines | -lines | Tokens |")
    out.append("|---|---|---|--:|--:|--:|")
    for (meta, added, removed) in rows:
        subj = meta["subject"].replace("|", "\\|")
        if len(subj) > 64:
            subj = subj[:61] + "..."
        tk = g(meta["tok"]["total"]) if meta["tok"]["has_data"] else "—"
        out.append(f"| {meta['date']} | `{meta['sha']}` | {subj} | {g(added)} | {g(removed)} | {tk} |")
    out.append("")
    return "\n".join(out)


def main() -> int:
    ap = argparse.ArgumentParser(description="OUTRAM PARK historian — pre-merge-to-main report")
    ap.add_argument("--from", dest="dfrom", help="window start DDMMYY (day-month-year, 2-digit year)")
    ap.add_argument("--to", dest="dto", help="window end DDMMYY (default: today)")
    ap.add_argument("--branch", default="develop", help="branch to report on (default: develop)")
    ap.add_argument("--base", default="main",
                    help="base branch for the default 'not yet in base' window (default: main)")
    ap.add_argument("--outfile", help="explicit output path (default: docs/historian/historian_<from>_to_<to>.md)")
    args = ap.parse_args()

    branch_ref = ref_for_branch(args.branch)

    dfrom = parse_ddmmyy(args.dfrom) if args.dfrom else None
    dto = parse_ddmmyy(args.dto) if args.dto else (_dt.date.today() if args.dfrom else None)
    base_ref = None
    if not args.dfrom:
        # Default window: commits on <branch> not yet on <base>.
        base_ref = ref_for_branch(args.base)

    commits = commits_in_range(branch_ref, base_ref, dfrom, dto)
    md = render(dfrom, dto, branch_ref, base_ref, commits)

    if args.outfile:
        outpath = args.outfile
    else:
        if dfrom and dto:
            tag = f"{dfrom.strftime('%d%m%y')}_to_{dto.strftime('%d%m%y')}"
        else:
            tag = f"since_{args.base}_to_{_dt.date.today().strftime('%d%m%y')}"
        outpath = os.path.join(HERE, f"historian_{tag}.md")

    os.makedirs(os.path.dirname(outpath), exist_ok=True)
    with open(outpath, "w", encoding="utf-8") as fh:
        fh.write(md)
    rel = os.path.relpath(outpath, repo_root())
    print(f"wrote {rel}  ({len(commits)} commits)")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: crates/kovan-metrics/src/historian.rs
//! The historian report — pre-merge-to-`main` accounting.
//!
//! Before `develop` is merged into `main`, the workspace generates a report
//! accounting for the **API tokens spent** and the **lines / KLOC written**
//! across the window of history being released.
//!
//! # Sources, not estimates
//!
//! - **Lines** come from `git log --numstat --no-merges` over the range.
//! - **Tokens** come from the `API-Usage-Since-Last-Commit` commit trailers.
//!
//! Commits predating the token hooks legitimately carry *no token data*. They
//! are counted in the line totals and shown with `—` in the token column. That
//! is correct output, not a gap to be filled in.
//!
//! # Default window
//!
//! With no `--from`, the window is "everything on `<branch>` not yet on
//! `<base>`" (i.e. `base..branch`), which is exactly what the pending merge
//! would deliver.

use std::path::{Path, PathBuf};

use crate::date::Date;
use crate::git::{self, NumStat};
use crate::trailer::{self, TokenCounts};

/// Where generated reports live, relative to the repository root.
pub const REPORT_DIR_REL: &str = "docs/historian";

/// Everything one commit contributes to the report.
struct Row {
    date: String,
    short: String,
    subject: String,
    stat: NumStat,
    tokens: TokenCounts,
    recorded_total: u64,
    has_token_data: bool,
}

/// Collect the commits in the window, newest-last, with their statistics.
fn collect(
    branch_ref: &str,
    base_ref: Option<&str>,
    from: Option<Date>,
    to: Option<Date>,
) -> Vec<Row> {
    let fmt = git::record_format();
    let since = from.map(|d| format!("--since={} 00:00:00", d.iso()));
    let until = to.map(|d| format!("--until={} 23:59:59", d.iso()));
    let range = match base_ref {
        Some(base) => format!("{base}..{branch_ref}"),
        None => branch_ref.to_string(),
    };

    let mut args: Vec<&str> = vec!["log", "--no-merges", "--reverse", &fmt];
    if let Some(s) = since.as_deref() {
        args.push(s);
    }
    if let Some(u) = until.as_deref() {
        args.push(u);
    }
    args.push(&range);

    git::parse_records(&git::git_output(&args))
        .into_iter()
        .map(|rec| {
            let parsed = trailer::parse(&rec.body);
            let (tokens, recorded_total, has_data) = match parsed {
                Some(t) => (t.counts, t.total, t.has_data()),
                None => (TokenCounts::default(), 0, false),
            };
            Row {
                stat: git::numstat(&rec.short),
                date: rec.date,
                short: rec.short,
                subject: rec.subject,
                tokens,
                recorded_total,
                has_token_data: has_data,
            }
        })
        .collect()
}

/// Render the report markdown.
fn render(
    from: Option<Date>,
    to: Option<Date>,
    branch_ref: &str,
    base_ref: Option<&str>,
    rows: &[Row],
) -> String {
    let g = trailer::group;
    let mut tok = TokenCounts::default();
    let mut tok_total = 0u64;
    let mut tok_covered = 0usize;
    let mut added = 0u64;
    let mut removed = 0u64;
    let mut rs_added = 0u64;
    let mut rs_removed = 0u64;
    let mut per_crate: Vec<(String, u64)> = Vec::new();

    for r in rows {
        tok.add(&r.tokens);
        tok_total += r.recorded_total;
        if r.has_token_data {
            tok_covered += 1;
        }
        added += r.stat.added;
        removed += r.stat.removed;
        rs_added += r.stat.rs_added;
        rs_removed += r.stat.rs_removed;
        for (name, n) in &r.stat.per_crate_added {
            match per_crate.iter_mut().find(|(k, _)| k == name) {
                Some((_, v)) => *v += n,
                None => per_crate.push((name.clone(), *n)),
            }
        }
    }

    let span = match (from, to, base_ref) {
        (Some(f), Some(t), _) => format!("{} → {}", f.human(), t.human()),
        (_, _, Some(base)) => format!("all of `{branch_ref}` not in `{base}`"),
        _ => format!("full history of `{branch_ref}`"),
    };
    let kloc = |n: u64| format!("{:.1}", n as f64 / 1000.0);
    let net = |a: u64, r: u64| a as i64 - r as i64;
    /// Thousands-group a possibly-negative net figure (a shrinking window is a
    /// legitimate result, so the sign must survive).
    fn group_i64(n: i64) -> String {
        let s = trailer::group(n.unsigned_abs());
        if n < 0 {
            format!("-{s}")
        } else {
            s
        }
    }
    let net_kloc = |a: u64, r: u64| format!("{:.1}", net(a, r) as f64 / 1000.0);

    let mut o = String::new();
    o.push_str(&format!("# OUTRAM PARK — historian report ({span})\n\n"));
    o.push_str(
        "> Pre-merge-to-`main` accounting of the API tokens spent and the lines / \
         KLOC written across this window of `develop` history. **Auto-generated** \
         by `kovan historian`; regenerate with \
         `kovan historian --from DDMMYY --to DDMMYY`.\n\n",
    );
    o.push_str("## Scope\n\n");
    o.push_str(&format!(
        "- **Branch:** `{branch_ref}`{}\n",
        base_ref.map(|b| format!(" (vs base `{b}`)")).unwrap_or_default()
    ));
    o.push_str(&format!("- **Window:** {span}\n"));
    o.push_str(&format!("- **Commits (non-merge):** {}\n", rows.len()));
    o.push_str(&format!(
        "- **Token coverage:** {tok_covered}/{} commits carry an \
         `API-Usage-Since-Last-Commit` trailer. Commits before the token-accounting \
         hooks existed (or made outside a Claude session) contribute 0 and are \
         counted here as *no token data* — that is correct, not missing data.\n\n",
        rows.len()
    ));

    o.push_str("## Totals\n\n");
    o.push_str("### Lines written (git numstat, merges excluded)\n\n");
    o.push_str("| Metric | Lines | KLOC |\n|---|--:|--:|\n");
    o.push_str(&format!("| Added (all files) | {} | {} |\n", g(added), kloc(added)));
    o.push_str(&format!("| Removed (all files) | {} | {} |\n", g(removed), kloc(removed)));
    o.push_str(&format!(
        "| **Net (all files)** | **{}** | **{}** |\n",
        group_i64(net(added, removed)),
        net_kloc(added, removed)
    ));
    o.push_str(&format!("| Added (Rust `.rs`) | {} | {} |\n", g(rs_added), kloc(rs_added)));
    o.push_str(&format!(
        "| Net (Rust `.rs`) | {} | {} |\n",
        group_i64(net(rs_added, rs_removed)),
        net_kloc(rs_added, rs_removed)
    ));

    o.push_str("\n### API tokens spent\n\n");
    o.push_str("| Component | Tokens |\n|---|--:|\n");
    o.push_str(&format!("| input | {} |\n", g(tok.input)));
    o.push_str(&format!("| output | {} |\n", g(tok.output)));
    o.push_str(&format!("| cache_read | {} |\n", g(tok.cache_read)));
    o.push_str(&format!("| cache_write | {} |\n", g(tok.cache_write)));
    o.push_str(&format!("| **total** | **{}** |\n", g(tok_total)));
    o.push_str(
        "\n_`total` = input + output + cache_read + cache_write. Cache-read \
         (prompt-cache re-reads of the growing context) usually dominates; the \
         output figure is the closest proxy for net generated content._\n\n",
    );

    if !per_crate.is_empty() {
        per_crate.sort_by(|a, b| b.1.cmp(&a.1).then(a.0.cmp(&b.0)));
        o.push_str("## Lines added, by crate (top 20)\n\n");
        o.push_str("| Crate | Lines added |\n|---|--:|\n");
        for (name, n) in per_crate.iter().take(20) {
            o.push_str(&format!("| `{name}` | {} |\n", g(*n)));
        }
        o.push('\n');
    }

    o.push_str("## Per-commit ledger\n\n");
    o.push_str("| Date | Commit | Subject | +lines | -lines | Tokens |\n");
    o.push_str("|---|---|---|--:|--:|--:|\n");
    for r in rows {
        let mut subj = r.subject.replace('|', "\\|");
        if subj.chars().count() > 64 {
            subj = subj.chars().take(61).collect::<String>() + "...";
        }
        let tk = if r.has_token_data {
            g(r.recorded_total)
        } else {
            "—".to_string()
        };
        o.push_str(&format!(
            "| {} | `{}` | {} | {} | {} | {} |\n",
            r.date,
            r.short,
            subj,
            g(r.stat.added),
            g(r.stat.removed),
            tk
        ));
    }
    o
}

/// Generate a historian report and write it to disk.
///
/// Returns the path written and the number of non-merge commits covered.
///
/// With `from` unset the window defaults to `base..branch` — everything on
/// `branch` not yet on `base` — and `to` defaults to today only when `from` was
/// given, matching the Python this replaces.
pub fn generate(
    from: Option<Date>,
    to: Option<Date>,
    branch: &str,
    base: &str,
    outfile: Option<PathBuf>,
) -> Result<(PathBuf, usize), String> {
    let branch_ref = git::ref_for_branch(branch);
    let to = match (from, to) {
        (Some(_), None) => Some(Date::today()),
        (_, t) => t,
    };
    let base_ref = if from.is_none() {
        Some(git::ref_for_branch(base))
    } else {
        None
    };

    let rows = collect(&branch_ref, base_ref.as_deref(), from, to);
    let md = render(from, to, &branch_ref, base_ref.as_deref(), &rows);

    let root = git::repo_root();
    let path = match outfile {
        Some(p) => p,
        None => {
            let tag = match (from, to) {
                (Some(f), Some(t)) => format!("{}_to_{}", f.ddmmyy(), t.ddmmyy()),
                _ => format!("since_{}_to_{}", base, Date::today().ddmmyy()),
            };
            root.join(REPORT_DIR_REL).join(format!("historian_{tag}.md"))
        }
    };
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)
            .map_err(|e| format!("cannot create {}: {e}", parent.display()))?;
    }
    std::fs::write(&path, md).map_err(|e| format!("cannot write {}: {e}", path.display()))?;

    let shown = path.strip_prefix(&root).unwrap_or(&path).to_path_buf();
    println!("wrote {}  ({} commits)", shown.display(), rows.len());
    Ok((path, rows.len()))
}

/// Resolve the default output path for a window, without writing anything.
pub fn default_output_path(root: &Path, from: Option<Date>, to: Option<Date>, base: &str) -> PathBuf {
    let tag = match (from, to) {
        (Some(f), Some(t)) => format!("{}_to_{}", f.ddmmyy(), t.ddmmyy()),
        _ => format!("since_{}_to_{}", base, Date::today().ddmmyy()),
    };
    root.join(REPORT_DIR_REL).join(format!("historian_{tag}.md"))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn row(date: &str, subject: &str, added: u64, removed: u64, tokens: Option<u64>) -> Row {
        Row {
            date: date.into(),
            short: "abc1234".into(),
            subject: subject.into(),
            stat: NumStat {
                added,
                removed,
                rs_added: added,
                rs_removed: removed,
                per_crate_added: vec![("tampines".into(), added)],
            },
            tokens: TokenCounts {
                input: tokens.unwrap_or(0),
                ..Default::default()
            },
            recorded_total: tokens.unwrap_or(0),
            has_token_data: tokens.is_some(),
        }
    }

    #[test]
    fn renders_totals_and_a_ledger() {
        let rows = vec![
            row("2026-08-12", "first", 100, 10, Some(5_000)),
            row("2026-08-13", "second", 250, 50, Some(7_500)),
        ];
        let md = render(
            Date::new(2026, 8, 1).ok(),
            Date::new(2026, 8, 13).ok(),
            "origin/develop",
            None,
            &rows,
        );
        assert!(md.contains("# OUTRAM PARK — historian report"));
        assert!(md.contains("01 Aug 2026 → 13 Aug 2026"));
        assert!(md.contains("**Commits (non-merge):** 2"));
        // 350/1000 formats as "0.3": 0.35 is not exactly representable and
        // lands just below the midpoint. Python's `f"{:.1f}"` agrees.
        assert!(md.contains("| Added (all files) | 350 | 0.3 |"));
        assert!(md.contains("**Net (all files)** | **290**"));
        assert!(md.contains("| Removed (all files) | 60 | 0.1 |"));
        assert!(md.contains("| **total** | **12,500** |"));
        assert!(md.contains("| `tampines` | 350 |"));
        assert!(md.contains("**Token coverage:** 2/2"));
    }

    #[test]
    fn commits_without_trailers_show_a_dash_not_a_zero() {
        // The honesty rule: absent data must be visibly absent.
        let rows = vec![row("2026-07-01", "pre-hooks commit", 40, 0, None)];
        let md = render(None, None, "origin/develop", Some("origin/main"), &rows);
        assert!(md.contains("| 40 | 0 | — |"), "missing token data must render as an em dash");
        assert!(md.contains("**Token coverage:** 0/1"));
        assert!(md.contains("all of `origin/develop` not in `origin/main`"));
    }

    #[test]
    fn a_pipe_in_a_subject_is_escaped_so_the_table_survives() {
        let rows = vec![row("2026-08-13", "fix a|b parsing", 1, 0, Some(1))];
        let md = render(None, None, "develop", None, &rows);
        assert!(md.contains("fix a\\|b parsing"));
    }

    #[test]
    fn long_subjects_are_truncated() {
        let long = "x".repeat(200);
        let rows = vec![row("2026-08-13", &long, 1, 0, Some(1))];
        let md = render(None, None, "develop", None, &rows);
        assert!(md.contains(&format!("{}...", "x".repeat(61))));
    }

    #[test]
    fn default_output_path_uses_the_ddmmyy_tag() {
        let p = default_output_path(
            Path::new("/repo"),
            Date::new(2026, 8, 1).ok(),
            Date::new(2026, 8, 13).ok(),
            "main",
        );
        assert!(p.ends_with("historian_010826_to_130826.md"));
    }
}
